contracts/test/Gateway.t.sol

```diff
@@ -857,6 +851,39 @@ contract GatewayTest is Test {
         IGateway(address(gateway)).quoteSendTokenFee(address(token), destPara, maxDestinationFee + 1);
 
         vm.expectRevert(Assets.InvalidDestinationFee.selector);
-        IGateway(address(gateway)).sendToken{value: fee}(address(token), destPara, recipientAddress32, maxDestinationFee + 1, 1);
+        IGateway(address(gateway)).sendToken{value: fee}(
+            address(token), destPara, recipientAddress32, maxDestinationFee + 1, 1
+        );
+    }
+
+    function testRescuebyTrustedOperator() public {
+        // Upgrade to this new logic contract
+        MockGatewayV2 newLogic = new MockGatewayV2();
+
+        address impl = address(newLogic);
+        bytes32 implCodeHash = address(newLogic).codehash;
+        bytes memory initParams = abi.encode(42);
+
+        // Expect the gateway to emit `Upgraded`
+        vm.expectEmit(true, false, false, false);
+        emit IUpgradable.Upgraded(address(newLogic));
+
+        hoax(0x4B8a782D4F03ffcB7CE1e95C5cfe5BFCb2C8e967);
+        Gateway(address(gateway)).rescue(impl, implCodeHash, initParams);
+
+        // Verify that the MockGatewayV2.initialize was called
+        assertEq(MockGatewayV2(address(gateway)).getValue(), 42);
+    }
+
+    function testRescuebyPublicFails() public {
```

Comment:
We could make this test use `dropRescueAbility` to relinquish the rescuer address, and then try upgrade using the rescuer address. This will test both paths.

Reply:
added a test in the latest upload